src/shared/Core/Trace2.cs

```diff
@@ -0,0 +1,316 @@
+
+    /// <summary>
+    /// Writes GCM/EVENT format version information as a message to the trace writer.
+    /// </summary>
+    /// <param name="gcmVersion">The version of GCM.</param>
+    /// <param name="eventFormatVersion">
+    /// The EVENT format version, which is defined in Git.
+    /// This may need to be incremented in the future to align with updated versions in Git.
+    /// </param>
+    /// <param name="filePath">Path of the file this method is called from.</param>
+    /// <param name="lineNumber">Line number of file this method is called from.</param>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1026:DefaultParametersShouldNotBeUsed")]
+    void WriteVersion(
+        string gcmVersion,
+        string eventFormatVersion = "3",
+        [System.Runtime.CompilerServices.CallerFilePath] string filePath = "",
+        [System.Runtime.CompilerServices.CallerLineNumber] int lineNumber = 0);
+
+    /// <summary>
+    /// Writes argv and time information associated with application start to
+    /// trace writer.
+    /// </summary>
+    /// <param name="argv">Exe name and arguments.</param>
+    /// <param name="elapsedTime">Time since application started running.</param>
+    /// <param name="filePath">Path of the file this method is called from.</param>
+    /// <param name="lineNumber">Line number of file this method is called from.</param>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1026:DefaultParametersShouldNotBeUsed")]
+    void WriteStart(
+        long elapsedTime,
+        [System.Runtime.CompilerServices.CallerFilePath] string filePath = "",
+        [System.Runtime.CompilerServices.CallerLineNumber] int lineNumber = 0);
+    
+    /// <summary>
+    /// Writes exit code and execution time to trace writer.
+    /// </summary>
+    /// <param name="code">Application exit code.</param>
+    /// <param name="timeInMilliseconds">Elapsed time in milliseconds since GCM started.</param>
+    /// <param name="filePath">Path of the file this method is called from.</param>
+    /// <param name="lineNumber">Line number of file this method is called from.</param>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1026:DefaultParametersShouldNotBeUsed")]
+    void WriteExit(
+        int code,
+        long elapsedTime,
+        [System.Runtime.CompilerServices.CallerFilePath]
+        string filePath = "",
+        [System.Runtime.CompilerServices.CallerLineNumber]
+        int lineNumber = 0);
+}
+
+public class Trace2 : DisposableObject, ITrace2
+{
+    private string _sid;
+    private string[] _args;
+    private CollectorWriter _writer;
+    private IEnvironment _environment;
+    
+    private const string GitSidVariable = "GIT_TRACE2_PARENT_SID";
+    
+    public long ApplicationStartTime { get; }
+
+    public Trace2(IEnvironment environment, string[] args, long applicationStartTime = 0)
```

Comment:
(_I'm having a hard time jumping from caller to callee in the PR view, so apologies in advance._)
So I'm assuming that this ctor() is called directly/indirectly from Main().

Wouldn't it be better to have the Main() pass down a DateTime set to UtcNow (as observed in Main())
or just let the ctor() compute it.  In either case, store that absolute_start_time as a private field inside the Trace2 class.

Then have context.Trace2.WriteVersion() just used the stored absolute start time in the message.

Then have the event methods (context.Trace2.WriteExit() and context.Trace2.WriteStart()) NOT take an elapsed time argument, but rather have them call DateTime.UtcNow() and use that in the JSON Time field.

For the events that need to emit a "t_abs" / elapsed time field, let those WriteFOO() methods do the subtraction and hide all of this from the callers.

Reply:
Sure, that's fine. I went ahead and made the field in `Trace2.cs` private, set the time in `WriteVersion` to the value of the field, and updated the WriteFOO() methods to do the subtraction.